sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -79,11 +85,149 @@
  * containing the message from the {@link PAssert} that failed.
  */
-public class TestPipeline extends Pipeline {
+public class TestPipeline extends Pipeline implements TestRule {
+
+  private static class PipelineRunEnforcement {
+
+    protected boolean enableAutoRunIfMissing;
+    protected final Pipeline pipeline;
+    private boolean runInvoked;
+
+    private PipelineRunEnforcement(final Pipeline pipeline) {
+      this.pipeline = pipeline;
+    }
+
+    private void enableAutoRunIfMissing(final boolean enable) {
+      enableAutoRunIfMissing = enable;
+    }
+
+    protected void beforePipelineExecution() {
+      runInvoked = true;
+    }
+
+    protected void afterTestCompletion() {
+      if (!runInvoked && enableAutoRunIfMissing) {
+        pipeline.run().waitUntilFinish();
+      }
+    }
+  }
+
+  private static class PipelinePAssertEnforcement extends PipelineRunEnforcement {
+
+    private List<TransformHierarchy.Node> runVisitedNodes;
+
+    private static class PAssertDetector extends PipelineVisitor.Defaults {
+
+      boolean hasPAssert = false;
+
+      private boolean isPAssertNode(final TransformHierarchy.Node node) {
+        return
+            node.getTransform() instanceof PAssert.GroupThenAssert
+                || node.getTransform() instanceof PAssert.GroupThenAssertForSingleton
+                || node.getTransform() instanceof PAssert.OneSideInputAssert;
+      }
+
+      @Override
+      public void leaveCompositeTransform(final TransformHierarchy.Node node) {
+        hasPAssert = hasPAssert || isPAssertNode(node);
+      }
+    }
+
+    private static class NodeRecorder extends PipelineVisitor.Defaults {
+
+      private final List<TransformHierarchy.Node> visited = new LinkedList<>();
+
+      @Override
+      public void leaveCompositeTransform(final TransformHierarchy.Node node) {
+        visited.add(node);
+      }
+
+      @Override
+      public void visitPrimitiveTransform(final TransformHierarchy.Node node) {
+        visited.add(node);
+      }
+
+    }
+
+    private PipelinePAssertEnforcement(final TestPipeline pipeline) {
+      super(pipeline);
+    }
+
+    private boolean hasPAssert(final Pipeline pipeline) {
+      final PAssertDetector pAssertDetector = new PAssertDetector();
+      pipeline.traverseTopologically(pAssertDetector);
+      return pAssertDetector.hasPAssert;
+    }
+
+    private List<TransformHierarchy.Node> recordPipelineNodes(final Pipeline pipeline) {
+      final NodeRecorder nodeRecorder = new NodeRecorder();
+      pipeline.traverseTopologically(nodeRecorder);
+      return nodeRecorder.visited;
+    }
+
+    private void verifyPipelineRan() {
+      final List<TransformHierarchy.Node> pipelineNodes = recordPipelineNodes(pipeline);
+      if (runVisitedNodes != null && !runVisitedNodes.equals(pipelineNodes)) {
+        throw new AbandonedPTransformException("The pipeline contains abandoned PTransform(s)"
```

Comment:
Do you mind putting the `AbandonedPTransformException` bit from the `PAssertMissingException` bit?

I really really want the `AbandonedPTransformEnforcement` (making up the name) to be on by default, but would rather have the `PAssertMissingEnforcement` off by default.

See also my comment on the tests below.